compiler/rustc_codegen_llvm/src/attributes.rs

```diff
@@ -335,6 +336,14 @@ pub fn from_fn_attrs<'ll, 'tcx>(
     to_add.extend(probestack_attr(cx));
     to_add.extend(stackprotector_attr(cx));
 
+    // When `no_builtins` is applied at the crate level, we should add the
+    // `no-builtins` attribute to each function to ensure it takes effect in LTO.
+    let crate_attrs = cx.tcx.hir().attrs(rustc_hir::CRATE_HIR_ID);
```

Comment:
Should this be based on the crate in which we codegen or based on the crate which defined the function? I think the later makes more sense as otherwise it would be possible for rustc to delay codegen of a function in a `#![no_builtins]` crate to when compiling a non-`#![no_builtins]` crate. The former matches the current behavior though.

Reply:
Maybe add `no_builtins` to `CodegenFnAttrs`? That is where other codegen affecting attributes like `#[linkage_name]` or `#[target_feature(enable = "...")]` are put.

Reply:
> Should this be based on the crate in which we codegen or based on the crate which defined the function? I think the later makes more sense as otherwise it would be possible for rustc to delay codegen of a function in a `#![no_builtins]` crate to when compiling a non-`#![no_builtins]` crate. The former matches the current behavior though.

Sorry, I didn't fully understand. Could you please clarify the specific differences between the two crates you're referring to?
I would like to add the `no-builtins` attribute to all functions in a crate that has `#![no_builtins]`. I attempted to add a test case for this case. I also tried it locally using `cargo`, and it seems to be consistent with the expected behavior.

Reply:
> Maybe add `no_builtins` to `CodegenFnAttrs`? That is where other codegen affecting attributes like `#[linkage_name]` or `#[target_feature(enable = "...")]` are put.

Yes, that's what I mentioned earlier. However, I'm not sure if it requires further discussion, such as considering additional options like `#[no_builtins(memcpy)]` for more flexible control.
If we simply add a `#[no_builtins]`, would you prefer it to be included in this PR or start a new one?

Reply:
> Could you please clarify the specific differences between the two crates you're referring to?

If a function is marked as `#[inline]` or is generic and used in a different crate the current logic will apply the `#![no_builtins]` attribute of the crate we are currently compiling rather than the crate in which the function was originally defined. The crate we are currently compiling likely doesn't have `#![no_builtins]`, thus causing the function from the `#![no_builtins]` crate to be incorrectly codegened without `#![no_builtins]`. By taking the `#![no_builtins]` attribute from the crate that originally defined the function we are currently compiling it would correctly get `no-builtins` applied.

Reply:
> Yes, that's what I mentioned earlier.

In this instance I meant adding the crate level `#![no_builtins]` to `CodegenFnAttrs`. If function level `#[no_builtins]` is added if either attribute is used the `no_builtins` field could be set to `true`. I think function level `no-builtins` makes sense, but it is indeed likely a discussion for a future PR.

Reply:
I added a `CodegenFnAttrFlags::NO_BUILTIN`, and it seems to export `no-builtins` correctly.